### PR TITLE
feat: add inlay hints for generic parameters

### LINKED ISSUE #11091: inlay-hints for generic arguments
We should offer the ability to show inlay hints for generic arguments, just like we do for function arguments already.
```rs
fn foo<'a, T, const C: usize>() {}
foo::<'b, u32, 0>()
// should render as something like
foo::<'a: 'b, T: u32, C: 0>()
```
Each of these 3 should have a config flag, with only the const generics one being enabled by default I'd say.

### PR DESCRIPTION
**Pull request by MikeWalrus**
fixes #11091

By default, only hints for const generic parameters are shown, and this can be configured through `rust-analyzer.inlayHints.genericParameterHints.enable`.

Probably needs more testing.

### CODE REVIEW

File: crates/ide/src/inlay_hints/param_name/generic.rs [hunk 1]
```diff
@@ -0,0 +1,288 @@
...
+
+use crate::{inlay_hints::GenericParameterHints, InlayHint, InlayHintsConfig, InlayKind};
+
+use super::{is_argument_similar_to_param_name, render_label};
+
+pub(crate) fn hints(
+    acc: &mut Vec<InlayHint>,
+    sema: &hir::Semantics<'_, RootDatabase>,
+    config: &InlayHintsConfig,
+    node: SyntaxNode,
```
Comment (Veykril) [inline crates/ide/src/inlay_hints/param_name/generic.rs:16]:
```suggestion
    node: AnyHasGenericArgs,
```
[`c08d419` (#17555)](https://github.com/rust-lang/rust-analyzer/pull/17555/commits/c08d419fba1c7ad3c6e80f38d9412942c8516b9a) adds that, we should use it here (will need a rebase)

File: crates/ide/src/inlay_hints/param_name/generic.rs [hunk 2]
```diff
@@ -0,0 +1,288 @@
...
+                    let const_param = ast::ConstParam::cast(syntax)?;
+                    const_param.name()?.syntax().clone()
+                }
+                hir::GenericParam::LifetimeParam(it) => {
+                    if const_generics_only {
+                        return None;
+                    }
+                    sema.source(it)?.value.syntax().clone()
+                }
+            };
```
Comment (Veykril) [inline crates/ide/src/inlay_hints/param_name/generic.rs:65]: Querying the source can be expensive, so lets move this down to where we actually need it so we give the other filtering things a run first

File: crates/ide/src/inlay_hints/param_name/generic.rs [hunk 3]
```diff
@@ -0,0 +1,288 @@
...
+
+            if should_hide {
+                return None;
+            }
+
+            let linked_location = sema.original_range_opt(&source_syntax);
+            let label = render_label(param_name, config, linked_location);
+            Some(InlayHint {
+                range: sema.original_range_opt(arg.syntax())?.range,
```
Comment (Veykril) [inline crates/ide/src/inlay_hints/param_name/generic.rs:82]: Likewise lets move this in front of the `let source_syntax` (after that was moved down) as its a cheaper filter

File: crates/ide/src/inlay_hints/param_name/generic.rs [hunk 4]
```diff
@@ -0,0 +1,288 @@
...
+                position: crate::InlayHintPosition::Before,
+                pad_left: false,
+                pad_right: true,
+                kind: InlayKind::Parameter,
```
Comment (Veykril) [inline crates/ide/src/inlay_hints/param_name/generic.rs:86]: Let's add a new one for this, GenericParameter` and map that to `PARAMETER`

**Review by Veykril — COMMENTED** [review]
Generally looks pretty good!

**Review by Veykril — COMMENTED** [review]
A test for when the user supplied mismatching args wrt to the params might be nice as well

**Comment by MikeWalrus** [comment]
@Veykril I've pushed again. Please let me know if I miss anything. :)

File: crates/rust-analyzer/src/config.rs [hunk 5]
```diff
@@ -2874,6 +2885,19 @@ fn field_props(field: &str, ty: &str, doc: &[&str], default: &str) -> serde_json
...
+                "always",
+                "never",
+                "const_only"
+            ],
+            "enumDescriptions": [
+                "Always show generic parameter hints.",
+                "Never show generic parameter hints.",
+                "Only show const generic parameter hints."
+            ]
+        },
```
Comment (Veykril) [inline crates/rust-analyzer/src/config.rs:2900]:
```suggestion
```

**Comment by Veykril** [comment]
Looks good, thanks!
@bors r+

**Pull request merged**